Add option to specify binmode, bump version.

brainbuz opened the pull request:
I encountered problems with characters above 128 and added an option to specify the binmode.

brainbuz committed:
Tie::File -- add binmode option, add unicode to pod, related issue #17494 (not fixed)

haarg commented on dist/Tie-File/t/44_binmode.t:
```diff
@@ -0,0 +1,46 @@
+use Test::More;
+
+use utf8;
+use feature 'unicode_strings';
```
This feature doesn't appear to be needed in this test. For now I think it could be removed.

haarg commented:
Could you update the commit message to note that this is a change in Tie-File?

Grinnz commented on dist/Tie-File/lib/Tie/File.pm:
```diff
@@ -2128,6 +2131,16 @@ then the array C<@gifts> would appear to contain C<("Gold\n",
+  use utf8;
+  use feature 'unicode_strings';
+  # tie a file with utf8 binmode
+  tie my @array, 'Tie::File', 'array.file', 'binmode' => ':utf8';
```
Please suggest ':encoding(UTF-8)' rather than ':utf8' particularly for reading files as the :utf8 layer can result in corrupted strings.

```suggestion
  tie my @array, 'Tie::File', 'array.file', 'binmode' => ':encoding(UTF-8)';
```

Grinnz commented:
The porting tests indicate dist/Tie-File/t/44_binmode.t must be added to the MANIFEST. I'm not sure if there is an automated tool for this but it should be straightforward.

jkeenan commented:
Assuming this p.r. is intended to fix https://github.com/Perl/perl5/issues/21867, please add that to the commit message.  Thanks.

brainbuz commented:
I switched the pod to :encoding(UTF-8), updated the commit message, also deleted the line require 5.005.
The issue James mentioned applies to a different module. I did find a related issue #17494 that I missed last night when I was looking for open issues, unfortunately this PR does not fix it, in fact when I made pointed the code that I submitted this for I was hit by the issue from 17494. I'll look at that issue and update this PR if I can fix it, if I can't fix it I would recommend adding a note to the doc on Tie::File that it should be considered read only on unicode files.

jkeenan commented on dist/Tie-File/lib/Tie/File.pm:
```diff
@@ -2128,6 +2129,16 @@ then the array C<@gifts> would appear to contain C<("Gold\n",
+
+  # make sure Perl is expecting unicode files and strings
+  use utf8;
+  use feature 'unicode_strings';
```
Your documentation for the `binmode` option states that to use it you have to ensure that the program "is expecting unicode file and strings", and that this is ensured by calling `use utf8; use feature 'unicode_strings';`.  However, the test program you've added to the repository, does not call `use feature 'unicode_strings';`.  Which is correct?  The documentation or the test program?

brainbuz replied:
Those 2 lines are recommended, one of the earlier comments requested I remove unicode_strings from the test because it isn't necessary for the test to work. I can make the doc clearer that these two lines are recommended.

jkeenan commented on dist/Tie-File/lib/Tie/File.pm:
```diff
@@ -2128,6 +2129,16 @@ then the array C<@gifts> would appear to contain C<("Gold\n",
+  use utf8;
+  use feature 'unicode_strings';
+  # tie a file with UTF-8 binmode
+  tie my @array, 'Tie::File', 'array.file', 'binmode' => ':encoding(UTF-8)';
```
One of our porting tests, `t/porting/podcheck.t`, imposes a 78-character line length in documentation and calculates tab expansion as part of that process.  This line is slightly over that calculated maximum.  A simple way to fix that would be to break the line in two at an appropriate point, *e.g.,*:
```
diff --git a/dist/Tie-File/lib/Tie/File.pm b/dist/Tie-File/lib/Tie/File.pm
index 9b41efa08a..b45aea9685 100644
--- a/dist/Tie-File/lib/Tie/File.pm
+++ b/dist/Tie-File/lib/Tie/File.pm
@@ -2137,7 +2137,8 @@ Specify the binmode to set for files opened by Tie::File.
   use utf8;
   use feature 'unicode_strings';
   # tie a file with UTF-8 binmode
-  tie my @array, 'Tie::File', 'array.file', 'binmode' => ':encoding(UTF-8)';
+  tie my @array,
+    'Tie::File', 'array.file', 'binmode' => ':encoding(UTF-8)';
 
 =head2 C<mode>
 
```
Here, too, run `make test_porting` before re-pushing.  Thanks.

jkeenan replied:
Also, I suggest changing the Subject of this GH pull request to be: "Tie-File: add option to specify binmode".  You can trim the first line of the commit message to simply that as well.

brainbuz replied:
Once I delved into the open issue #17494, I determined that this pull request only works for reading files and have work in another branch https://github.com/brainbuz/perl5/tree/trytofix17494, I was able to improve behavior but writes of new data are still 1 character off.
I'd like to wait a few days to see if I can solve this problem, if I can't I'll make the minor changes and add to the doc a warning that with non-byte files Tie::File should be used read only. If I can fix it I'll merge it to this pull request.

Tux replied:
I see *no* added value in adding "recommended" `use` statements in documentation if there is no relation to the stuff being documented.
`use utf8;` is about the source code and has no relation whatsoever to the file being tied, so it distracts from what is documented. My first though is "why do I need this?"
I'd say leave out what is not required.

jkeenan reviewed (changes requested):
Our "porting" tests impose QA standards and have to be placated before our GH CI programs can run to completion.  Please adjust as needed.  This will facilitate evaluation of the actual code changes you're proposing.  Thanks.

brainbuz commented:
I'm closing this and submitting a new PR for just a documentation patch. I tried to fix 17494, without that it isn't safe to write to unicode file handles with Tie::File.

brainbuz commented:
I'm closing this and submitting a new PR for just a documentation patch. I tried to fix 17494, without that it isn't safe to write to unicode file handles with Tie::File.
